pkg/rtc/user_packet_deduper.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2023 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rtc
+
+import (
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/livekit/protocol/livekit"
+)
+
+const (
+	maxSize = 100
+)
+
+type UserPacketDeduper struct {
+	lock sync.Mutex
+	seen map[uuid.UUID]uuid.UUID
+	head uuid.UUID
+	tail uuid.UUID
+}
+
+func NewUserPacketDeduper() *UserPacketDeduper {
+	return &UserPacketDeduper{
+		seen: make(map[uuid.UUID]uuid.UUID),
+	}
+}
+
+func (u *UserPacketDeduper) IsDuplicate(up *livekit.UserPacket) bool {
+	u.lock.Lock()
```

Comment:
we don't need to acquire the lock if there's no nonce

Reply:
Good catch. Will move it.

Reply:
Changed in this commit - https://github.com/livekit/livekit/pull/3366/commits/6470cda79668dd00696512daa8cf538593daf69d